### DISCUSSION TITLE
Improving the docs

### DISCUSSION THREAD

**Discussion by jghauser** [papis/papis#413] [Ideas]
I'm thinking about rewriting/cleaning up the documentation. I think this has been something that users have often commented on (there are a lot of `faq` and `documentation` issues). I have some experience making technical information accessible and was thinking of giving this a stab.

What's the best way of testing docs generation locally? Is there a auto-formatter for the `.rst` files (I see they are hardwrapped, what are the settings)?

After gathering some ideas, I will draft a more complete outline of the overall structure and ask for feedback on that. Maybe we can then start a new branch to work on this (@alexfikl said he'd be interested in helping).

Happy to hear any feedback you may have, both when it comes to ideas for improvement and how to best structure the rewriting process. A bunch of things below end with a '?', that usually meant I wasn't sure about it. I would be particularly happy about feedback on that :)

Sorry for the big mess of notes!

# Readme (and other files in the main folder)

- get rid of all or most the banners ;)
  - it looks messy and i'm not sure it adds much
  - which ones to keep?
- i would make it much shorter
  - the quick start thing should just point to the docs
- it should have an actual description (rather than pointing to various other places)
- rename 'contact' to 'support'
  - get rid of zulip (we use the discussions, and I think it's good to try to minimise channels of communications but keep those active)
- have a section 'mentions'
  - we can put all the blog posts etc mentioning papis in there
- move some `.md` and `.rst` files to readthedocs (have one place for all docs except readme)
  - CHANGELOG? (maybe keep)
  - CODE_OF_CONDUCT (maybe keep because it's dev oriented?)
  - CONTRIBUTING (maybe keep (as it's dev and not user oriented)
    - or have a developer section in the readthedocs?
  - HACKING
    - i think we could have a "advanced" section in the readthedocs where we put this (more on this below)
  - INSTALL (is there any reason for this? if you happen to only have the cloned repo, you should find the relevant info/links in the readme)
  - ROADMAP. convert to issues? (could tag them to show they are roadmap)
- papis-libgen actually links to papis-firefox
- is papis-scihub still functional?
- is papis-firefox still functional?
- is papis-zotero still functional?
  - i think when i tried using it ages ago, sqlite import somehow didnt work i think (but it's been a long time)
  - and i think bibtex import can be done by `papis` now
  - i think it would be best if we have one main suggested path of migration (also considering that now there is the zotero2papis script as well)

# Readthedocs

- migrate to version two of readthedocs?
- remove sections
  - scihub support?
  - anything else?
- maybe a new overall structure (I'm thinking the top level sections to be like in the readthedocs docs, so just dividers). Maybe the following organisation:

## 1) Getting started

- quick guide
- workflow and common tasks (#328, #304)
- installation options
  - move the "running tests" stuff to CONTRIBUTING.md
- import (i think this should be introduced quite earlier, because it will be crucial for almost all new users)
  - what is the recommended approach here?
    - there is papis-zotero
    - but #281 says we can just use `papis bibtex`

## 2) Installation

- pretty much as is, but simplifying things a bit
- maybe remove the "sudo pip" options? I think that's generally not recommended and people who do want it probably already know about it?

## 3) Configuration

- are they all in the example config? (I think not, or at least I think I remember once finding some options in the source that I didn't know about)
- also some could use better explanations
  - `add-file-name` and `add-folder-name` (see #373)

## 4) Core concepts

- includes the following existing sections
  - info.yaml
  - library structure
  - database

## 5) Commands

- add missing commands
- "Main" command
  - is confusingly named (at least we could say that this is just `papis`)
  - should be first in list
- commands should all be named `papis [command]`
- generally make the sections for each command less confusing
  - I find it confusing that there's first e.g. "Add" then "Cli" then "papis add"
    - can we change this? or is this required given how the docs are generated from comments in the source files?
  - less subheadings (it's hard to tell at what level you are)
  - maybe better structure would be
    1) short explanation (1-2 sentences)
    2) longer description (1 paragraph)
    3) options
    4) examples
- improve explanations
  - bibtex (see #339)

## 6) Interfaces

- includes the following sections
  - Native
    - improve documentation (see #367)
  - Fzf
  - Web
  - Editor integration
  - Launchers (hmm, is there a better name for this?)
    - Rofi (is this still working?)
    - Dmenu (is this still working?)

## 8) Advanced features

- includes the following existing sections
  - config.py
  - scripting
    - scripts should be either in the docs (or wiki, see below) or the example/scripts folder
    - i think this file here should maybe focus on explaning *how to* script
    - and refer to the examples/scripts folder for the relevant examples (rather than having them twice)
  - api
  - plugin architecture
  - git
  - hooks
  - document

## 9) FAQ

- add things from existing `faq` issues
  - field/tag removal with `papis update` (#358)
  - transfer to table (#336)
  - get dois/axiv ids from pdfs (#332)
    - or should that be in some command? @alejandrogallo 
  - empty bibtex entries (#407)
  - batch processing (#250)
  - auto update of entries (#313)
- move shell-autocompletion here
  - improve (see #303 and #399)
- i think we can close faq issue #327 (user stopped responding and I don't understand what it's about)

# Wiki

- maybe create something like user scripts/configs section?
- or should we put that into the readthedocs?
  - pro wiki: this is more of a wiki think (user contributions, possibly a bit messy, collaboratively edited)
  - pro readthedocs: only one source of information
- i have some things users might be interested in, mostly some pretty advanced jinja2 settings and a sexy fzf setup
- maybe best to remove wiki and put evrything into readthedocs (just to have one source only)

# Other random thoughts that came to me while working on this

- I think I might be able to get us a logo as well (nothing fancy, but nice enough)
  - do you have any suggestions?
  - i think it should probably include a book or so and then something techy/cli-y :laughing: 
- while at it i was thinking also to clean up the repo a bit
  - docker stuff working or useful?
  - if so, we should document it somewhere
- maybe make a simple website (though I don't think it's that important)

**Comment by alexfikl** [comment]
Overall this looks really nice and thought out to me! :tada:

> What's the best way of testing docs generation locally? Is there a auto-formatter for the .rst files (I see they are hardwrapped, what are the settings)?

I'll give it a better read and add some comments here and there later. For this part though, you can generate the docs with
```bash
cd doc
make html SPHINXOPTS="-W --keep-going -n"
xdg-open build/html/index.html
```
or just open them with you favorite browser. I haven't touched the `papis` docs much, but the only rule seems to be to keep the lines at a width of 80-ish. Do we need any others? Maybe worth looking into something like https://github.com/sphinx-contrib/sphinx-lint?

**Reply by jghauser** [reply]
Ok great, thanks! I just figured out how to get the hard wrap working in neovim, never used that before (I use softwrap in my prose files). I'll check out the linter, thanks! That should be useful even if we don't really need any other formatting rules (can't think of anything off the top of my head).

**Comment by alexfikl** [comment]
Hi! Went through some of your suggestions / questions for the `README` and left some random thoughts. Hope it helps!

> # Readme (and other files in the main folder) 
>     * get rid of all or most the banners ;)

Some of those banners are really common for open source projects, e.g. CI, docs, pypi. The only one that truly bothers me is the `Packaging Status` one which ruins the layout.

I would say that they're all ok to have there, because they show a quick glance at the state / activity of the project. Maybe we could organize them into columns instead of rows or something? So that they look less cluttered.
 
>     * i would make it much shorter
>       * the quick start thing should just point to the docs

There's probably room for improvement, but something like what's already in the README seems like a good idea. When first discovering the project, it was really helpful to see a few screenshots and some examples of the command-line commands to get a feel for it right away.

There's a lot more of that in the docs (which is definitely a better place for it), but leaving some in the README seems ok too for that quick impact. Not sure what a good length would be though..

>     * rename 'contact' to 'support'
>       * get rid of zulip (we use the discussions, and I think it's good to try to minimise channels of communications but keep those active)

This sounds good! Can't say I've ever used any of the support channels, so don't have a strong sense of which are actually useful.

>     * have a section 'mentions'
>       * we can put all the blog posts etc mentioning papis in there

There's some of that in the `Description` section. We can break it off into its own subsection or something? 

>     * move some `.md` and `.rst` files to readthedocs (have one place for all docs except readme)
>       * CHANGELOG? (maybe keep)
>       * CODE_OF_CONDUCT (maybe keep because it's dev oriented?)
>       * CONTRIBUTING (maybe keep (as it's dev and not user oriented)
>         * or have a developer section in the readthedocs?
>       * HACKING
>         * i think we could have a "advanced" section in the readthedocs where we put this (more on this below)
>       * INSTALL (is there any reason for this? if you happen to only have the cloned repo, you should find the relevant info/links in the readme)
>       * ROADMAP. convert to issues? (could tag them to show they are roadmap)

Most of these are pretty standard open source project files to have. Not quite sure if they're more of an "unspoken rule" though, since I don't remember seeing anything written down. We can probably take a look at some other modernized packages to see what they do these days?

I definitely agree it's better to move the ROADMAP to Issues or Projects on Github, since that's where all that discussion takes places anyway.

>     * papis-libgen actually links to papis-firefox
>     * is papis-scihub still functional?
>     * is papis-firefox still functional?
>     * is papis-zotero still functional?
>       * i think when i tried using it ages ago, sqlite import somehow didnt work i think (but it's been a long time)
>       * and i think bibtex import can be done by `papis` now
>       * i think it would be best if we have one main suggested path of migration (also considering that now there is the zotero2papis script as well)

Some of those seem to still have some maintainers, but not sure if they still work :(
Would say we can just leave them there for publicity, i.e. maybe they'll find some contributors in time.

For migration, there's already
https://papis.readthedocs.io/en/latest/importing.html
We should clean that up and put it more front and center somewhere? 

Personally, I migrated from Mendeley with the bibtex import and it worked very well. That also seems like the best way to do it because the bibtex support has a good chance of being maintained long term.

**Comment by alexfikl** [comment]
Left a couple of comments in the `readthedocs` section as well. Most of it seems very nicely thought  out!

> # Readthedocs
> 
>     * migrate to version two of readthedocs?

In progress at https://github.com/papis/papis/pull/414!

>     * maybe a new overall structure (I'm thinking the top level sections to be like in the readthedocs docs, so just dividers). Maybe the following organisation:

Very different in scope probably, but I've found the `jax` documentation to be very well organized into "tutorials", "technical details" and "API reference manual"
https://jax.readthedocs.io/en/latest/

> ## 1) Getting started
>     * import (i think this should be introduced quite earlier, because it will be crucial for almost all new users)
>       * what is the recommended approach here?
>         * there is papis-zotero
>         * but [The readme page uses the option `--from-doi`, which doesn't exist #281](https://github.com/papis/papis/issues/281) says we can just use `papis bibtex`

Agreed this should be higher up somewhere! I'm not sure if there currently is a recommended way to do the import, so maybe best would be just to list a few? We can also point to the `papis` simple `yml` file format and make it clear it's very hackable, i.e. one can easily write their own "my_format_to_papis" script.

> ## 2) Installation
>     * maybe remove the "sudo pip" options? I think that's generally not recommended and people who do want it probably already know about it?

:+1: and some mentions of `pip3` in there, which are quite outdated!

> ## 3) Configuration
>     * are they all in the example config? (I think not, or at least I think I remember once finding some options in the source that I didn't know about)

Need to check these are still in sync, definitely! Would be nice if there was some way to automate that..

> ## 5) Commands
>     * generally make the sections for each command less confusing
>       * I find it confusing that there's first e.g. "Add" then "Cli" then "papis add"
>         * can we change this? or is this required given how the docs are generated from comments in the source files?

We should be able to clean that up somehow. It seems like that `.. click` directive in
https://github.com/papis/papis/blob/master/papis/commands/add.py#L85-L88
adds an extra heading that we would need to remove or modify to just say `Cli`. I can look into that!

**Comment by alexfikl** [comment]
> I'm thinking about rewriting/cleaning up the documentation. I think this has been something that users have often commented on (there are a lot of faq and documentation issues). I have some experience making technical information accessible and was thinking of giving this a stab.
>
> What's the best way of testing docs generation locally? Is there a auto-formatter for the .rst files (I see they are hardwrapped, what are the settings)?
>
> After gathering some ideas, I will draft a more complete outline of the overall structure and ask for feedback on that. Maybe we can then start a new branch to work on this (@alexfikl said he'd be interested in helping).
>
> Happy to hear any feedback you may have, both when it comes to ideas for improvement and how to best structure the rewriting process. A bunch of things below end with a '?', that usually meant I wasn't sure about it. I would be particularly happy about feedback on that :)
>
> Sorry for the big mess of notes!

I managed to go through all of it and it would be awesome if we implemented this! `papis` has a lot of cool functionality all over the place and is quite easily extensible, so some improved documentation should do wonders!

For how to start actually working on this, I would say that making a branch and working there has a big potential of becoming outdated or just overwhelming. A lot of the changes you proposed can be done piece by piece in small-ish chunks, so I would recommend we go that route instead and improve incrementally each section of the docs.

**Reply by jghauser** [reply]
Thanks a lot for your feedback! Those are all very useful suggestions! I'm thinking as a first step I'm gonna look at rewriting the README as it's arguably the most important part of the docs. I'm gonna write a PR and implement my and your suggestions and then we can discuss it further there (you're definitely right about splitting the work up into many chunks!). I'm gonna take a look at some other projects regarding the banners that we might want to keep. I do really want to limit them to two rows maximum though ;).

Btw, why does the CI banner say "failing"? Is that related to the tests you've been working on?

**Reply by alexfikl** [reply]
> Btw, why does the CI banner say "failing"? Is that related to the tests you've been working on?

Yep, some of the tests started failing because of some changes and new dependencies. #403 should fix that and, hopefully, make it less likely to happen in the future!

**Comment by alejandrogallo** [comment]
Thank you both for these thoughts.
The documentation of papis has certainly become a bit of a mess
and some parts are outdated, that's a pitty.

I agree with most that @jghauser has pointed out
and up to now it has worked great with the Readme.

I don't know what is the best way of going about updating the
documentation.
Perhaps we could organize a jitsi video meeting to discuss this
more efficiently and I think that'd be nice too.

- The Wiki part really resonated with me and I went ahead
  and created an example of how this could be:
    https://github.com/papis/papis/wiki/User-scripts-examples
  I think that'd be great if people can write there stuff.
- I also think the Zulip experiment kind of failed, I also think that
  the github discussions are more than enough, I just have the fear
  that we can't export the discussions in the future, but I guess it's
  possible, I have to look into it.
- I also think we should really deprecate packages that don't work anymore
  or really look for people that use them to actively maintain them.
  It was too much work for me to update everything in every release.
- I also think the commands are a mess, the only relevant information is in the
  beginning and the explanations are sometimes a word salad.
  I think we should have a user manual and a developer manual, and all the
  API stuff from the commands should go to the developer manual, far away
  from the users' eyes.

Great work!

**Reply by jghauser** [reply]
Thanks for the feedback! I think a jitsi meeting would be nice, we could also use that to discuss the fosdem idea you mentioned in the other discussion. :)

**Comment by jghauser** [comment]
Some thoughts re organisation of command help from #432 :

This is much better. I have some other ideas for improvement. I'll use the `add` command to illustrate.

1. Multiple descriptions of what commands do.

There are sort of two short descriptions of what the commands do. One is at the very top of the help "The add command..." and the second is after "Command line interface", where it says "Add a document...". I think the two should be integrated. I think we might want to delete the description in the comment at the top of `add.py` and add the description to the `click` stuff at the bottom of the file. In that way the more complete description will appear also when using `--help`.

2. Order of information

I think we can delete the "command-line interface" heading and move all information under that heading to the very beginning. This documention is about the CLI commands and I think we don't need a subheading with that title.

Maybe the order could be something like:
- info coming from click help
- examples
- a ref to relevant configuration options (in the second paragraph of the `add` documentation, there's a reference to such options, and I think that might be some info that shouldn't be shown with `--help` (as would happen if we do (1)). Or maybe we just delete this entirely, but make it easily findable in the config section of the docs.

3. Information that shouldn't be there

Stuff in `Examples in python` shouldn't be here, but in some "Advanced config" section or so.

You can also just merge this and I can implement my suggestions myself in a follow-up commit (if others think this makes sense).

**Comment by alejandrogallo** [comment]
Quick question, do you think it's meaningful to use markdown?
I hate rst veven more that markdown and it's always quite a pain to use.
I don't know if other people also feel like it, and I think to accept documentation
contributions these days people prefer markdown. And apparently markdown is supported
https://www.sphinx-doc.org/en/master/usage/markdown.html

I would prefer of course org-mode, but this will enrage some people ;)

**Reply by alexfikl** [reply]
I've been writing rst for too long now thanks to sphinx, so I don't feel strongly either way :))
But I do agree that org-mode would be the nicest of the three :heart:

**Reply by alexfikl** [reply]
Does markdown (with sphinx, right?) support all the LaTeX stuff?

**Reply by jghauser** [reply]
I've actually been thinking about this, but I'm not sure it's worth the effort given that everything's in rst so far. I'm more at ease with markdown, but found rst not hard to get into. And I'm not sure how easy/difficult it is to get the full sphinx functionality from markdown (I haven't investigated this). It would probably make future contributions easier as more people know markdown...

IF the full functionality can be had with markdown and IF conversion is straightforward, I'd have a slight preference for markdown.

Org-mode is cool and all, but the new and even cooler kid on the block is [neorg](https://github.com/nvim-neorg/neorg) ;)

**Reply by alejandrogallo** [reply]
I'll look into the markdown option, for the sake of future contributions I think it is the most sane and adult option ;) I won't pronounce myself on neorg... ;D

**Reply by alejandrogallo** [reply]
but neorg ist not compatible with org... I don't see an easy way of also respecting the `.. ` directives with markdown, with org-mode this is trivial however.
What I do in my personal projects is writing in org mode and then converting with emacs to rst
and building with sphinx, like here https://github.com/alejandrogallo/atrip
and results in this https://alejandrogallo.github.io/atrip/atrip.html.
But of course I don't know how well is the state of org-mode in vim, and apparently neorg is not backwards compatible with org-mode...

**Reply by alexfikl** [reply]
[vim-orgmode](https://github.com/jceb/vim-orgmode) tries to be compatible, from what I remember. It's still pretty far away from the nice native emacs experience though (at least it was a couple of years ago)..

**Reply by alexfikl** [reply]
Have you looked at https://myst-parser.readthedocs.io/en/latest? It's markdown-like with support for sphinx. Not sure the language extensions make it any easier to use than rst though.

**Reply by alejandrogallo** [reply]
Yes, I have been playing with myst, I guess I have to play more with, I'll see.
About vim-orgmode, yes, I have also a package I maintain for my colleagues i nordr to have only simple syntax highlighting in vim, because in the end this is all someone might care about, they don't get much more for rst anyways.

The thing is that now I have to write a lot of documentation, and I find myself everytime writing in org-mode and then exporting to rst, and it's kind of annoying and it gets more annoying with the passing of years. But I guess that should be ok.

**Reply by jghauser** [reply]
There's actually a very decent [org-mode clone for neovim](https://github.com/nvim-orgmode/orgmode). Neorg is an entirely new and different project, trying to modernise the orgmode file format. I really like where it's headed and am using it, but it's still in alpha. The file spec just about to reach 0.1 and much of the functionality only exists on paper so far. So my mention of it was mostly tongue-in-cheek. :)

Btw. I'm still very motivated to do a lot work in rewriting the docs. It's just that for personal reasons I don't have a lot of time for it until around mid January. So far I've mostly been thinking about what a sane overall structure would look like. I like how active papis has gotten recently, well done you two! :tada:

**Reply by alejandrogallo** [reply]
it's only thanks to you guys ;)
